Bluetooth: ISO: Remove TS size from SDU len check

Thalley opened the pull request:
The timestamp is not part of the SDU, and should
thus not be used to get the maximum SDU size.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

fixes https://github.com/zephyrproject-rtos/zephyr/issues/49629

Thalley commented:
Found a similar issue in multiple samples, as well as one of the babblesim tests. 

Seeing how we need to deal with the ISO header, I can't help but to wonder if we can improve this somehow. 

Basically with the current API we have
1) Reserve 4-9 bytes based on different factors 
2) Add SDU size - 4/8 octets based on the timestamp parameter
3) Send data 

I wonder if we can do this better, without having to resort to copying data?

cvinayak commented on samples/bluetooth/iso_broadcast/src/main.c:
```diff
@@ -44,7 +44,7 @@ static struct bt_iso_chan_ops iso_ops = {
 
 static struct bt_iso_chan_io_qos iso_tx_qos = {
-	.sdu = sizeof(uint32_t), /* bytes */
+	.sdu = BT_HCI_ISO_DATA_HDR_SIZE + (sizeof(uint32_t)), /* bytes */
```
This does not seems correct in terms of the definition of `sdu`:
`/** Channel SDU. Maximum value is BT_ISO_MAX_SDU */`

and

https://github.com/zephyrproject-rtos/zephyr/blob/b8cdf9d0a6372ce72727ca3dc4fcc39fc9ec66d6/include/zephyr/bluetooth/iso.h#L66-L67

Why cant the required change be done internally instead of each samples?

Thalley replied:
The ISO SDU contains the data header, so the 4/8 octets of data header is counted towards the SDU size. 

So the maximum SDU size is indeed 0xFFF, but the maximum data length in the SDU is 0xFFF - 4 or 8, depending on whether the timestamp is added. 

I would really like this to be handled internally, but I am unsure how to do that since the buffer where the SDU (including the timestamp) is allocated by the application.

Casper-Bonde-Bose commented on samples/bluetooth/broadcast_audio_source/src/main.c:
```diff
@@ -63,7 +63,8 @@ static void stream_sent_cb(struct bt_audio_stream *stream)
 	net_buf_reserve(buf, BT_ISO_CHAN_SEND_RESERVE);
-	net_buf_add_mem(buf, mock_data, preset_16_2_1.qos.sdu);
+	net_buf_add_mem(buf, mock_data,
+			preset_16_2_1.qos.sdu - BT_HCI_ISO_DATA_HDR_SIZE);
```
I think this is wrong. (Of cause this app is just sending mock data, but it sends the wrong message to people using the sample to build an application.
QoS sdu size is the negotiated over-the-air max SDU size - which do not carry any HCI headers - hence there should be no need to reserve space for the header in the payload area of buf. That would mean you would need to setup the ISO link to be able to carry 128 octets to fit a 120 byte audio frame - hence wasting 8 bytes of air-time for each frame.
The buffers in the host needs to be allocated large enough to hold both the ISO max SDU as well as the HCI header, but the data path should not care about the HCI header when building the SDU.

Thalley replied:
Will update this PR to properly fix the issue. The PSN and TS should not be considered part of the SDU

Thalley committed:
Bluetooth: ISO: Remove TS size from SDU len check

The timestamp is not part of the SDU, and should
thus not be used to get the maximum SDU size.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Casper-Bonde-Bose reviewed (dismissed):
LGTM

stephanosio commented:
Rebased to fix CI failure

Frodevan commented on subsys/bluetooth/host/iso.c:
```diff
@@ -723,24 +723,6 @@ static uint16_t iso_chan_max_data_len(const struct bt_iso_chan *chan,
-	 * the header size into account, as that also needs to be inserted into
-	 * the SDU
-	 */
-	if (ts == BT_ISO_TIMESTAMP_NONE) {
```
I seem to remember this was added to make sure a buffer was not overrun in an border-case EBQ test. Should the check be removed for both cases here, or just the case where ts != BT_ISO_TIMESTAMP_NONE?

Thalley replied:
The timestamp is not related to the SDU size (timestamp is part of the header, and not the SDU), so should just never have been added. 

If there's an issue with an EBQ tests without this, then the issue lies elsewhere

Thalley commented:
Updated the PR to also revert the (now wrong) documentation for the send functions

fabiobaltieri commented:
Merging since this was previously approved by the assignee.
